File: C++/740. Delete and Earn.cpp
class Solution {
public:
    int deleteAndEarn(vector<int>& nums) {

        if (nums.empty())
            return 0;

        const int len =  *max_element(nums.begin(), nums.end()); // the max num in nums used to be the length
        vector<int> value(len + 10, 0);
        //value store num and corresponding value
        for (auto &it : nums) {
            value[it] += it;
        }
        //dp store max_value before index i(include i)
        vector<int> dp(len + 10, 0);
        dp[0] = value[0];
        dp[1] = value[1];
        //calculate the current dp  in max(dp[1 position in front of current index],dp[2 position in front of current index] + value[current position])

        for (int i = 2; i < dp.size(); i++) {
            dp[i] = max(dp[i - 1], dp[i - 2] + value[i]);
        }

        return dp.back();
    }
};

File: Python/740. Delete and Earn.py
class Solution:
    def deleteAndEarn(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        # Dynamic Programming
        if len(nums) == 0:
            return 0

        num_count = collections.Counter(nums)
        # value store num and corresponding value
        value = collections.defaultdict(int)
        for num in num_count.elements():
            value[num] = num*num_count[num]

        # dp store max_value before index i(include i)
        dp = [0 for _ in range(max(nums)+10)]
        dp[0] = value[0]
        dp[1] = value[1]

        for i in range(2, max(nums)+10):
            # current value is the max between the value before and the value 2 position in front of current.   like  1-2 Stair climbing problem
            dp[i] = max(dp[i-2] + value[i], dp[i-1])

        return dp[-1]
